### ISSUE TITLE
STL Object Allocated in Stack in Lua Function That May Throw Lua Error May Leak Memory

### ISSUE THREAD

**Issue by MikuAuahDark** [love2d/love#1862]
Sorry for the bit confusing name, but the **TL;DR** is: If an STL (or non-POD) object is allocated in stack, and a Lua C API function is called, and that Lua C API function throws Lua error, the object destructor may not be called.

One example is in Lua function of `love.graphics.clear`

An `std::vector<OptionalColorf>` is allocated in stack.
https://github.com/love2d/love/blob/e5495d643662dbae0018e6daa170b13a5ca474e8/src/modules/graphics/wrap_Graphics.cpp#L76

However, later, `luaL_checknumber` is used. This function can throw Lua error if the specified index is not a number (or string that can't be converted to number). When the function is indeed throw Lua error, `std::vector<OptionalColorf>` destructor is not called, thus leaking memory.
https://github.com/love2d/love/blob/e5495d643662dbae0018e6daa170b13a5ca474e8/src/modules/graphics/wrap_Graphics.cpp#L109-L111

For a bit of background, Lua function that throws error (`lua_error`, `luaL_check*`, ...) calls [`longjmp`](https://en.cppreference.com/w/cpp/utility/program/longjmp). `longjmp` does NOT play well with C++ stack unwinding and this may result C++ destructors not being called.

The resulting memory leak can only be noticed if one of these is true:
* LOVE is compiled with Lua other than LuaJIT (except MSVC)
* LOVE is compiled with LuaJIT using MSVC (currently LOVE is not compiled with `/EHa`)
* LOVE is compiled with LuaJIT but with compiler-OS combination that result in [limited or no exception interopability](https://repo.or.cz/luajit-2.0.git/blob_plain/refs/heads/v2.1:/doc/extensions.html#exceptions)

This issue is not obvious enough because:
* MSVC internally performs stack unwinding when `longjmp` is called, however this is [compiler-specific](https://learn.microsoft.com/en-us/cpp/cpp/using-setjmp-longjmp?view=msvc-170) feature.
* LuaJIT tries to stack unwind when possible, as per exception interpoability table linked above.

Update: Apparently LuaJIT + MSVC is affected by this issue. LOVE needs to be compiled with `/EHa` for LuaJIT unwinding to work.

**Comment by MikuAuahDark** [comment]
Another example where this issue is present is in `love.graphics.captureScreenshot(string)`

First, `std::string` is allocated on stack.
https://github.com/love2d/love/blob/e5495d643662dbae0018e6daa170b13a5ca474e8/src/modules/graphics/wrap_Graphics.cpp#L516

However, if the image extension is unknown (either by providing extension other than PNG or TGA or not providing extension at all), then it throws Lua error through `luax_enumerror`.
https://github.com/love2d/love/blob/e5495d643662dbae0018e6daa170b13a5ca474e8/src/modules/graphics/wrap_Graphics.cpp#L527-L528

Thus, we can test it quickly by passing huge string to the `love.graphics.captureScreenshot` without any valid extension. So, to test if your installation is affected by this memory leak issue, use this Lua script.
```lua
local love = require("love")

local doError = ""

function love.update()
	collectgarbage("collect")
	if #doError > 0 then
		print(pcall(love.graphics.captureScreenshot, doError))
		doError = ""
	end
end

function love.draw()
	love.graphics.print("Press enter to leak memory.")
end

function love.keyreleased(k)
	if k == "return" then
		doError = string.rep("a", 1048576*100)
	end
end

local errhand = love.errorhandler or love.errhand

function love.errorhandler(msg)
	collectgarbage("collect")
	collectgarbage("collect")
	return errhand(msg)
end
```
If pressing enter increase the memory usage, then your OS-compiler combination is affected by this issue.

**Comment by slime73** [comment]
Yeah, Lua doesn't really give any better options here when working with C++ so historically we decided to stick with being able to actually write legible and maintainable binding code rather than trying to work around issues like this. In practice it hasn't really been a problem as far as I've been able to tell, which makes sense IMO because a Lua error is generally a problem to fix in a user's Lua code already (rather than hiding the problem via `pcall`).

Work that goes into #1640 might reduce the number of situations where this happens, and maybe it makes sense to figure out specific APIs that are very likely to be used with `pcall`  and make them safer to use that way. But aside from that I don't think solving this project-wide is worth the cost to code quality.
